### PR TITLE
Adds support for undici tracing

### LINKED ISSUE #1615: Support for undici tracing
Hi there!

We're starting to shift away from axios / request and wish to roll out undici in our new deployments.
Unfortunately, there is no 'official' tracing package for it.

Is it something that is being worked on? Would a PR be accepted to add an `undici` plugin? Should I be aware of any 🐉 ?

Thanks!

### PR DESCRIPTION
**Pull request by flovilmart**
### What does this PR do?

This PR adds support for tracing undici, for the upcoming release.

Closes #1615 

### Motivation

Undici does not use the standard HTTP / HTTPS modules from node.js. 
This is required so we can properly trace undici requests.

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [x] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

**Comment by rochdev** [comment]
> any suggestion regarding the node-ext and upstream koa failures? Is my setup wrong?

For upstream `koa` they are also failing on `master` so you can ignore them. For `next` I think rebasing might fix it since there was a recent change to how it's tested.

> the channels are weak ref'd and I've spent about 2 days trying to understand why my tracing would not work

We were also bitten by this while implementing `diagnostics_channel`. This is supposed to be an implementation details but is impacting common use cases enough that it should be changed. I'll open a PR to Node when I have some time.

**Comment by flovilmart** [comment]
Hi there! Any update on this one?

**Review by rochdev — CHANGES_REQUESTED** [review]
Sorry for the late reply as we've been trying to move the plugin system to `diagnostics_channel` and figuring out the details of how to properly use it ourselves.

It turns out that `undici` cannot be instrumented using DC in its current state because it doesn't have the proper channels/events for context propagation, which we need for spans to be connected together properly and for tracking the execution and control flows. This is something we'll have to work with @mcollina to support for both `undici` and `fastify` when we have finalized the design. The short version is that in order for `AsyncLocalStorage` to be used properly, there must be points at which to call `enterWith()` both at the beginning and synchronous end of a function for a client. Without this we cannot activate the span properly without leaking to the callback. So we need a minimum of 4 events (`start`, `end`, `async-end`, `error`) and `undici` only implements 3.

In the meantime, can you alter the integration to monkey-patch `undici` instead of using DC? It's not ideal, but unfortunately it's the only way short-term, and it will have the benefit of supporting older versions as well. At least most of the current logic should remain untouched other than the instrumentation points.

File: packages/datadog-plugin-undici/src/index.js [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+const HTTP_HEADERS = formats.HTTP_HEADERS
+const HTTP_STATUS_CODE = tags.HTTP_STATUS_CODE
+const HTTP_REQUEST_HEADERS = tags.HTTP_REQUEST_HEADERS
+const HTTP_RESPONSE_HEADERS = tags.HTTP_RESPONSE_HEADERS
+const SPAN_KIND = tags.SPAN_KIND
+const CLIENT = kinds.CLIENT
+const REFERENCE_CHILD_OF = opentracing.REFERENCE_CHILD_OF
+const REFERENCE_NOOP = constants.REFERENCE_NOOP
+
+function addError (span, error) {
```
[the remark this answers is not in the document]
Comment (flovilmart) [reply packages/datadog-plugin-undici/src/index.js:24]: perfect! Let me have a look at how I can get the implementation going for the monkey patch as well.

**Comment by mcollina** [comment]
> callback. So we need a minimum of 4 events (start, end, async-end, error) and undici only implements 3.

I do not have the time to look into it right now, but adding the missing one is just a PR away. It would probably take less time and be more maintenance friendly to just do that.

**Comment by flovilmart** [comment]
@rochdev , if I recap, in order to use diagnostics_channels, we need to implement the event handling through async storage.
We cannot currently properly implement async_storage as we're missing an event from undici, (`end` or `async-end` I assume).
@mcollina or @rochdev  do you happen to know where this 4th event need to be emitted from? I really wish to move forwards with diags channels, as the monkey patch implementation is not trivial coming from there.

**Comment by rochdev** [comment]
@flovilmart  The events that are needed are basically:

* **start:** When the operation starts, which would usually be when the function from the public API is called. We need this event to create a span, create a new `AsyncLocalStorage` store, and add the span to the store.
* **end:** When the operation ends but on the current stack, so not the entire asynchronous operation but only the synchronous part of it, which would usually be when the function from the public API returns. We need this event to restore the previous store from before the function was called.
* **error:** When any error occurs, which is needed to attach the error to the current span.
* **async-end:** When the operation finished completely, including anything asynchronous that it was waiting on. We need this to finish the span.

Additionally, while promises are handled automatically by `async_hooks`, callbacks are not, so any callback needs to be properly bound to the correct `AsyncResource`.

So an example fake `request` function could look something like this:

```js
const { AsyncResource } = require('async_hooks')

function request (callback) {
  const req = {}

  callback = AsyncResource.bind(callback)

  startChannel.publish(req)

  doActualAsyncRequest(req, (err, res) => {
    if (err) {
      errorChannel.publish(err)
    }

    asyncEndChannel.publish(req)

    callback(req, res)
  })

  endChannel.publish(req)
}
```

**Comment by flovilmart** [comment]
Oh great! Thanks @rochdev ; let me see how I can get this implemented.

**Comment by flovilmart** [comment]
@rochdev , I'm curious; and perhaps the current implementation is completely wrong, but given that this implementation keeps a WeakMap of `Request -> Spans`, isn't it circumventing this problem by manually creating / terminating spans? 
I'm currently running this implementation in a production service, and the traces are sane 🤔

**Comment by rochdev** [comment]
@flovilmart The context is never activated in the current implementation, unlike [http](https://github.com/DataDog/dd-trace-js/blob/2ca997c93838eba0c24ccb7a0bc171c43d41d10b/packages/datadog-plugin-http/src/client.js#L76), so any child of `undici` would end up in a different context. Since this is a client, it's less likely to have children, and when it does it could be less often, for example `tcp.connect`, which is why it appears to be working properly for your use case. However, for tracing we shouldn't assume what can or cannot have children so this should be handled. The problem is that without the missing event, handling it would be impossible since you would have no way to know when the function returns, meaning that the new context holding the span would leak after the function call. The same issue would potentially affect the callback if it's not properly bound with `async_hooks`.

Unfortunately these are issues that are much easier to solve with monkey-patching, but since `undici` is in active development it should also be relatively straightforward to add the missing events/bindings to the library itself and then use that in this PR which would only require small changes.

We're also working on a doc to explain how to do tracing using DC, especially since it's much more involved to get it right compared to monkey-patching, which makes it more difficult to use for external contributors.

**Comment by flovilmart** [comment]
Thanks a lot @rochdev for taking the time for this explanation! I concur that monkey patching is more straightforward then! Let me reevaluate both approaches.

**Comment by rochdev** [comment]
@flovilmart Sorry for the runaround with this PR. We hadn't quite realized that the current events were incomplete when we recommended trying to use DC for this. Let me know if you need any help to bring this to the finish line or have any questions about either approach.

**Comment by flovilmart** [comment]
@rochdev  no problem! All's good; I did not realize we were missing so many events.

**Comment by artur-ma** [comment]
@rochdev 

> async-end: When the operation finished completely, including anything asynchronous that it was waiting on. We need this to finish the span.

Does it mean the "async-end" should be emitted when body is consumed (read end from socket) ?
If so, @flovilmart correct me if im wrong the current implementation with WeakMap is wrong, since the user is receiving "Readable" instance which may be consumed at any time later, means the span does not include the socket read time.

**Comment by artur-ma** [comment]
@rochdev 

> async-end: When the operation finished completely, including anything asynchronous that it was waiting on. We need this to finish the span.

Does it mean the "async-end" should be emitted when body is consumed (read end from socket) ?
If so, @flovilmart correct me if im wrong the current implementation with WeakMap is wrong, since the user is receiving "Readable" instance which may be consumed at any time later, means the span does not include the socket read time.

**Comment by rochdev** [comment]
@artur-ma According to the current implementation for `http` yes, it would be after the response has been received. However, these are just the bare minimum required for tracing to work at all, but it definitely wouldn't hurt to have more events that are more precise, for example adding some sort of `response` or `headers-received` or something similar, which I believe already exist in Undici although I haven't looked for a few weeks.

**Comment by flovilmart** [comment]
@artur-ma sure, this doesn't include body reading probably it would need it's own span outside of the request span. What do you think?

**Comment by artur-ma** [comment]
AFAIK, anyway even if the user will not use the body, it should be consumer or destroyed by design
https://github.com/nodejs/undici#garbage-collection

I think, its possible to add such event on destroy / consume

**Comment by flovilmart** [comment]
@artur-ma that seems reasonable. I'm still on the fence about the monkey patching vs intrumentation with channels. I haven't had the chance to go back to it.

**Comment by mcollina** [comment]
> The only downside I see here, is that the user may choose not to consume it immediately but after some user action, which will be included in overall request time

I think that's correct and maps what the user is doing, i.e. it's something I would want to see in the traces.

**Comment by flovilmart** [comment]
With regards to the recent announcement that undici's fetch would be integrated in node as experimental. I believe it makes sense I wrap first the fetch implementation through patching, without diagnostic channels. Then I'll try to get something done in a more generic manner; what do you think?

**Comment by mcollina** [comment]
I still recommend getting this right. It will take a few months before this is unflagged.

**Comment by flovilmart** [comment]
Gotcha!

**Comment by lucasmarcelli** [comment]
Would this PR allow distributed tracing to connect services that are making requests via undici? We recently changed to undici for many of our server-to-server requests and the services are no longer connected on their traces. Does anyone know a workaround to connect them in the meantime? Are there headers I can set on the undici requests?

**Comment by rochdev** [comment]
@lucasmarcelli You can inject the current tracing context in the headers like this:

```js
const headers = {}
const span = tracer.scope().active()

tracer.inject(span, 'http_headers', headers)

client.dispatch({
  path: '/',
  method: 'GET',
  headers
}
```

**Comment by lucasmarcelli** [comment]
@rochdev thank you so much, i will give that a try :)

File: packages/datadog-plugin-undici/src/index.js [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        arguments[2] = AsyncResource.bind(arguments[2])
+      }
+      const scope = tracer.scope()
+      const childOf = scope.active()
+      const span = tracer.startSpan('http.request', {
+        childOf,
+        tags: {
+          [SPAN_KIND]: CLIENT
+        }
+      })
```
Comment (flovilmart) [inline packages/datadog-plugin-undici/src/index.js:316]: we init the span early and pass it through the asyncLocalStorage. not sure this is exactly what was intended

File: packages/datadog-plugin-undici/src/index.js [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    const span = asyncLocalStorage.getStore()
+    if (span) {
+      span.addTags({
+        'resource.name': method,
+        'span.type': 'http',
+        'http.method': method,
+        'http.url': uri,
+        'service.name': getServiceName(tracer, config, request)
+      })
+      requestSpansMap.set(request, span)
```
Comment (flovilmart) [inline packages/datadog-plugin-undici/src/index.js:98]: we keep the request in the WeakMap as for some reason, they're not available in the subsequent diagnostics channels events. Which makes me wonder why? and also if we can rely on this span to exist in the `undici:request:create` event

**Comment by flovilmart** [comment]
@rochdev  I believe I made some strides in the right direction:

- calls are now wrapped
- we use diagnostics channels in order to
  - inject span with all the request info
  - inject propagation headers in the request
  - capture connect / abort errors
  - capture status code
- we use the wrapping to
  - initially create the spans
  - propagate the async context
  - properly wrap the `pipeline` command
  - finish the spans when the promise resolves
  
 I also extracted commons from the `http` client wrapper as suggested.
 The test suite covers both using `client.[method]` and `undici.[method]` and ensure both emit traces properly at a high level.
 
❓  If we want to be extensive, we should run the whole suite with all methods we'll probably find some wholes in there.

 The instrumentation is a bit all over the place with the 2 wrapping strategies, I could attempt to simplify a bit 😅.

let me know what you think.

File: packages/datadog-plugin-undici/src/index.js (hunk 3 above)
Comment (mcollina) [reply packages/datadog-plugin-undici/src/index.js:93]: Can you open an issue about this? Or a PR to add them? saving the WeakMap will greatly reduce overhead.

File: packages/datadog-plugin-undici/src/index.js [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+  const statusCode = res.status || res.statusCode
+  span.setTag(HTTP_STATUS_CODE, statusCode)
+
+  if (!config.validateStatus(statusCode)) {
+    span.setTag('error', 1)
+  }
+}
+
+function addResponseHeaders (res, span, config) {
+  config.headers.forEach((key) => {
```
Comment (mcollina) [inline packages/datadog-plugin-undici/src/index.js:177]: A for...of is significantly faster.

Comment (flovilmart) [reply packages/datadog-plugin-undici/src/index.js:177]: Thanks for the tip! Does a shared resource exist for such optimisations as an Eslint rule set for example?

File: packages/datadog-plugin-undici/src/index.js (hunk 3 above)
Comment (flovilmart) [reply packages/datadog-plugin-undici/src/index.js:94]: Not sure what you mean. Is that more efficient to store through the weak map than the async context? 
or would you want something on the request object to add external context in the form of a weak map?

**Comment by mcollina** [comment]
This is a bit tricky and I don't have the solution at the top of my head without digging deep into the code.

What I would try first is to add one event listeners for the `'close'` event on the stream body itself.

**Comment by flovilmart** [comment]
Thanks for the tip @mcollina !

**Comment by flovilmart** [comment]
closing as another implementation is coming.

**Comment by ezequielmiranda87** [comment]
Hey, there @flovilmart  Just wanted to check in on this. Any updates on the new implementation? Is it in progress, or do you need any help? Thanks.

**Pull request closed without merging**